torch/csrc/Exceptions.cpp

```diff
@@ -177,54 +177,58 @@ PyWarningHandler::PyWarningHandler() noexcept(true):
 PyWarningHandler::~PyWarningHandler() noexcept(false) {
   c10::Warning::set_warning_handler(prev_handler_);
 
-  if(warning_buffer_.size() > 0) {
-    if(in_exception_) {
-      // An error happened after the warning
-      // Simply handle with the previous handler
-      for(const auto& warning: warning_buffer_) {
-        auto source_location = warning.source_location_;
-        const auto& msg = processErrorMsg(warning.msg_);
-        c10::Warning::warn(source_location, msg, warning.verbatim_);
+  if (warning_buffer_.size() > 0) {
+    PyObject *type, *value, *traceback;
+    pybind11::gil_scoped_acquire gil;
```

Comment:
Now we actually acquire the gil when an exception is being thrown. Is that ok to do?

Reply:
I would expect the GIL to be released when we reach this, in the exception case just like in the warning case (my understanding is that destructors for RIIA-style GIL will have been called).

Reply:
> I would expect the GIL to be released when we reach this

I wouldn't be so sure. All the `THPVariable` uses this while holding the GIL. And it is not released in the middle in most cases. So that wouldn't be a problem.
Things like the autograd engine do the same but release the GIL in the middle. But it will be reacquired when an exception occurs and so its not an issue (you can re-take the GIL without any issues with this API).
What I'm wondering is for things that call it without holding the GIL already. But I don't remember where these come from.

We were having issue before when the original warning call was trying to get the GIL. But this is in the main thread so should be good :)

Reply:
:sweat_smile: